deployment/scripts/upgrade.sh

```diff
@@ -0,0 +1,375 @@
+    waitForStop $pod "component=kibana"
+  done
+
+  local elasticsearch_dc=(`oc get dc -l logging-infra=elasticsearch -o jsonpath='{.items[*].metadata.name}'`)
+  for dc in "${elasticsearch_dc[@]}"; do
+    oc scale --replicas=0 dc $dc
+  done
+  for pod in $(oc get pods -l component=es -o jsonpath='{.items[*].metadata.name}'); do
+    waitForStop $pod "component=es"
+  done
+}
+
+function scaleUp() {
+
+  # scale up ES slowly -- we want to wait for each instance to recover its indices first
+  local elasticsearch_dc=(`oc get dc -l logging-infra=elasticsearch -o jsonpath='{.items[*].metadata.name}'`)
+  for dc in "${elasticsearch_dc[@]}"; do
+    oc scale --replicas=1 dc $dc
+
+    pod=$(getPodName $dc)
+    waitForStart $pod
+
+    # the first one we start gets to be the master
+    checkESStarted $pod
+  done
+
+  local kibana_dc=(`oc get dc -l logging-infra=kibana -o jsonpath='{.items[*].metadata.name}'`)
+  for dc in "${kibana_dc[@]}"; do
+    oc scale --replicas=1 dc $dc
+
+    pod=$(getPodName $dc)
+    waitForStart $pod
+    checkKibanaStarted $pod
+  done
+}
+
+function updateImages() {
+
+  # create any missing imagestreams and then update them all
+  oc new-app logging-imagestream-template || : # these may fail if created independently; that's ok
+
+  # TODO: update these to fail if we can't import?
+  oc import-image logging-elasticsearch || :
+  oc import-image logging-fluentd || :
+  oc import-image logging-kibana || :
+  oc import-image logging-curator || :
+}
+
+# we can just specify to regenerate our certs
+function add_curator() {
+
+  echo "Adding curator..."
+  # this may fail otherwise if there was only a partial upgrade previously
+  oc delete dc,rc,template --selector logging-infra=curator || :
+
+  generate_curator_template
+  generate_curator
+}
+
+function fluentd_daemonset() {
+  echo "Updating Fluentd to use Daemonset..."
+
+  # this may fail otherwise if there was only a partial upgrade previously
+  oc delete dc,rc,template,daemonset --selector logging-infra=fluentd || :
+
+  generate_fluentd_template
+  generate_fluentd
+}
+
+function upgrade_notify() {
+  set +x
+  echo "Upgrade complete!"
+  cat <<EOF
+
+=================================
+
+The deployer has created additional secrets, templates, and component deployments
+as part of the upgrade process. You may now have a few more steps to run manually.
+Consult the deployer docs for more detail.
+
+Fluentd:
+--------------
+Fluentd is deployed to nodes via a DaemonSet. If this is new for your deployment:
+as a cluster admin label the nodes you'd like to deploy it to:
+    oc label node/<node-name> ${fluentd_nodeselector}
+
+To label all nodes at once:
+    oc label nodes --all ${fluentd_nodeselector}
+
+Note: if your previous deployment used a DaemonSet for Fluentd, there should be
+no additional actions to deploy your pods -- the deployer did not unlabel any nodes.
+EOF
+}
+
+function regenerate_secrets() {
```

Comment:
yeah, i'll split up the logic within generate_secrets to not also create the support objects...